src/PostgREST/Config.hs

```diff
@@ -502,19 +501,37 @@ readPGRSTEnvironment =
 -- "postgres:///postgres?host=server&port=5432&fallback_application_name=PostgREST%2011%271%260%40%23%24%25%2C.%3A%22%5B%5D%7B%7D%3F%2B%5E%28%29%3Dasdfqwer"
 --
 -- >>> addFallbackAppName ver "postgres://user:invalid_chars[]#@host:5432/postgres"
--- "postgres://user:invalid_chars[]#@host:5432/postgres"
+-- "postgres://user:invalid_chars[]#@host:5432/postgres?fallback_application_name=PostgREST%2011.1.0%20%285a04ec7%29"
 --
--- >>> addFallbackAppName ver "invalid_uri1=val1 invalid_uri2=val2"
--- "invalid_uri1=val1 invalid_uri2=val2"
+-- >>> addFallbackAppName ver "host=localhost port=5432 dbname=postgres"
+-- "host=localhost port=5432 dbname=postgres fallback_application_name='PostgREST 11.1.0 (5a04ec7)'"
+--
+-- >>> addFallbackAppName strangeVer "host=localhost port=5432 dbname=postgres"
+-- "host=localhost port=5432 dbname=postgres fallback_application_name='PostgREST 11\\'1&0@#$%,.:\"[]{}?+^()=asdfqwer'"
 addFallbackAppName :: ByteString -> Text -> Text
 addFallbackAppName version dbUri = dbUri <>
-  case uriQuery <$> parseURI (toS dbUri) of
-    -- Does not add the application name to key=val connection strings or invalid URIs
+  case pgConnString dbUri of
     Nothing  -> mempty
-    Just ""  -> "?" <> uriFmt
-    Just "?" -> uriFmt
-    _        -> "&" <> uriFmt
+    Just PGKeyVal -> " " <> keyValFmt
+    Just PGURI    -> case T.breakOn "?" dbUri of
+      (_, "")  -> "?" <> uriFmt
+      (_, "?") -> uriFmt
+      (_, _)   -> "&" <> uriFmt
```

Comment:
On PostgREST 12.0.2 an URI without a percent encoded `?` works OK:

```
db-uri = "postgres://authenticator:pass?special@localhost:5432/postgres"
```

But, unless it's percent encoded, it gives this error with this change:

```
FATAL:  database \"postgres&fallback_application_name=PostgREST 12.1 (557d008)\" does not exist\n"
```

I think it still needs to check that the `breakOn` is done after the first appearance of `/`?  (after the `uriDesignator` of course)

Reply:
Good catch. Fixed that.

Reply:
Turns out that logic wasn't enough in cases like this:

```
"postgresql://?dbname=postgres&host=/run/user/1000/postgrest/postgrest-with-postgresql-16-BuR/socket&user=some_protected_user&password=invalid_pass"
```

That's currently mangled to:
```
"postgresql://?dbname=postgres&host=/run/user/1000/postgrest/postgrest-with-postgresql-16-BuR/socket&user=some_protected_user&password=invalid_pass?fallback_application_name=PostgREST%201
1.1.0%20%285a04ec7%29"
```